public/app/plugins/datasource/grafana-azure-monitor-datasource/components/MetricsQueryEditor/dataHooks.ts

```diff
@@ -0,0 +1,268 @@
+
+  return useAsyncState(
+    async () => {
+      if (!(subscription && resourceGroup && metricDefinition && resourceName && metricNamespace)) {
+        return;
+      }
+
+      const results = await datasource.getMetricNames(
+        subscription,
+        resourceGroup,
+        metricDefinition,
+        resourceName,
+        metricNamespace
+      );
+
+      const options = results.map(toOption);
+
+      if (metricName && !hasOption(options, metricName)) {
+        onChange(setMetricName(query, undefined));
+      }
+
+      return options;
+    },
+    setError,
+    [subscription, resourceGroup, metricDefinition, resourceName, metricNamespace]
+  );
+};
+
+export const useMetricMetadata = (query: AzureMonitorQuery, datasource: Datasource, onChange: OnChangeFn) => {
+  const [metricMetadata, setMetricMetadata] = useState<MetricMetadata>({
+    aggOptions: [],
+    timeGrains: [],
+    dimensions: [],
+    isLoading: false,
+    supportedAggTypes: [],
+    primaryAggType: undefined,
+  });
+
+  const { subscription } = query;
+  const { resourceGroup, metricDefinition, resourceName, metricNamespace, metricName, aggregation, timeGrain } =
+    query.azureMonitor ?? {};
+
+  // Fetch new metric metadata when the fields change
+  useEffect(() => {
+    if (!(subscription && resourceGroup && metricDefinition && resourceName && metricNamespace && metricName)) {
+      return;
+    }
+
+    datasource
+      .getMetricMetadata(subscription, resourceGroup, metricDefinition, resourceName, metricNamespace, metricName)
+      .then((metadata) => {
+        // TODO: Move the aggregationTypes and timeGrain defaults into `getMetricMetadata`
+        const aggregations = (metadata.supportedAggTypes || [metadata.primaryAggType]).map((v) => ({
+          label: v,
+          value: v,
+        }));
+
+        setMetricMetadata({
+          aggOptions: aggregations,
+          timeGrains: metadata.supportedTimeGrains,
+          dimensions: metadata.dimensions,
+          isLoading: false,
+          supportedAggTypes: metadata.supportedAggTypes ?? [],
+          primaryAggType: metadata.primaryAggType,
+        });
+      });
+  }, [datasource, subscription, resourceGroup, metricDefinition, resourceName, metricNamespace, metricName]);
+
+  // Update the query state in response to the meta data changing
+  useEffect(() => {
+    const aggregationIsValid = aggregation && metricMetadata.supportedAggTypes.includes(aggregation);
+
+    const newAggregation = aggregationIsValid ? aggregation : metricMetadata.primaryAggType;
+    const newTimeGrain = timeGrain || 'auto';
+
+    if (newAggregation !== aggregation || newTimeGrain !== timeGrain) {
+      onChange({
+        ...query,
+        azureMonitor: {
+          ...query.azureMonitor,
+          aggregation: newAggregation,
+          timeGrain: newTimeGrain,
+        },
+      });
+    }
+  }, [onChange, metricMetadata, aggregation, timeGrain, query]);
+
+  return metricMetadata;
+};
```

Comment:
are these functions something that will stay once this is migrated to the resource picker? If so, I'd add tests. (This apply to any new code added in this PR :) )

Reply:
great idea!